docs/slicing.md

```diff
@@ -0,0 +1,143 @@
+# Slicing
+
+Most sets in arrai can be sliced with the slice expression.
+
+The slicing expression can be used by putting the expression as an argument to a call.
+
+The syntax of the slicing expression is as follows:
+
+```text
+start=expr? ";" end=expr? (";" step=expr)?
+```
+
+As shown above, you can define a slice by defining a lower bound, an upper bound, both or none.
+You can also define a `step` that is used to define the increment of the numbers from the lower
+bound to the upper bound. When `step` is not defined, its value defaults to `1`.
+
+## Usages
+
+Only certain types of sets can be used with the slicing expression. These sets are:
+
+1. `Array`
+2. `String`
+3. `Byte Array`
+4. `Dictionary`
+
+Generic sets (i.e. unindexed sets) or any other expression that does not evaluates to the
+special types of sets will fail the expression.
+
+Example usages:
+
+| expression | equals |
+|:-|:-|
+| `[1, 2, 3, 4, 5](1;4)` | `[2, 3, 4]` |
+| `"this is a sentence"(;7)` | `"this is"` |
+| `{1: "hello", 2: "there", 3: 123, 4: 456}(;4)` | `["hello", "there", 123]` |
+| `"sentence"(;)` | `"sentence"` |
+| `[1, 2, 3, 4, 5](1;5;2)` | `[2, 4]` |
+| `[1, 2, 3, 4, 5](3;0;-1)` | `[4, 3, 2]` |
+| `[1, 2, 3, 4, 5](;1;-1)` | `[5, 4, 3]` |
+| `[1, 2, 3, 4, 5](3;;-1)` | `[4, 3, 2, 1]` |
+| `[1, 2, 3, 4, 5](;;-1)` | `[5, 4, 3, 2, 1]` |
+
+Since the expression is also an argument, the following expression is also valid
+
+```text
+[1, 2, 3, 4, 5](;;-1, 0)
+```
+
+this expression reverses the array and fetch the first value. In
+this example, it will evaluate to `5`.
+
+### Behaviour
+
+#### Default values
+
+Slice expression behaves differently when provided with different expressions.
+
+When the lower bound, the upper bound, and step are provided, the values will have
+to evaluate to a number as slicing can only be done with number expressions.
+Anything other than `Number`, the expression will fail.
+
+When `step` is not defined, its value defaults to `1`. If `step` evaluates to `0`,
+it will return an empty set. The value of `step` determines the default value of
+`start` and `end`.
+
+When `start` or `end` are defined, the values will be used. However, when the range
+is invalid (i.e `start > end && step > 0` or `start < end && step < 0`), an empty
+set will be returned as the result of the slice expression.
+
+When `start` or `end` are not defined, the value of `step` will determine their
+values.
+
+If `step` is positive, `start` will have the lowest possible value. In
+`Array` and `String`, the value of `start` defaults to the offset value. In
```

Comment:
it basically offsets the indexes, this might give a better explanation https://github.com/arr-ai/arrai/blob/master/docs/intro.md#some-lexical-conventions

Reply:
Ah, I see, maybe a link to that in the docs could help?